Admin Router: Conditional request buffering for /service endpoint

vespian opened the pull request:
## High-level description

This PR makes it possible to configure `/service` endpoint request buffering basing on the `DCOS_SERVICE_REQUEST_BUFFERING` label. See the Jira issue for details.

## Corresponding DC/OS tickets (obligatory)

These DC/OS JIRA ticket(s) must be updated (ideally closed) in the moment this PR lands:

  - https://jira.mesosphere.com/browse/DCOS-20269 `Request entity too large error`

## Related PRs

EE PR: https://github.com/mesosphere/dcos-enterprise/pull/2587

vespian commented:
@mesosphere-mergebot bump-ee

vespian commented:
@mesosphere-mergebot bump-ee

vespian commented:
@mesosphere-mergebot bump-ee

vespian commented:
@mesosphere-mergebot bump-ee

vespian commented:
@mesosphere-mergebot bump-ee

vespian commented:
@mesosphere-mergebot label Ready for review

vespian commented:
@sambatyon @jgehrcke @adamtheturtle  Please have a look

jgehrcke commented on packages/adminrouter/extra/src/README.md:
````diff
@@ -521,6 +521,48 @@ It is possible to workaround it by using another another HTTP reverse proxy in t
+```json
+{
+  "labels": {
+    "DCOS_SERVICE_REQUEST_BUFFERING": false
````
GitHub's highlighting shows that this lacks a comma :)

jgehrcke commented on packages/adminrouter/extra/src/includes/server/master.conf:
```diff
@@ -172,32 +172,57 @@ location ~ ^/service/(?<serviceid>[0-9a-zA-Z-.]+)$ {
+        -- labels. The proxy_request_buffering directive by itself is unable to
+        -- parse lua variables directly.
+        if service.request_buffering_required(service_name, marathon_cache) == false then
+            return ngx.exec("@service_requnbuffered");
```
Happy that we got this working!!

Maybe add a comment here saying

```
# Do an internal redirect; proceed handling this request in the 
# @service_requnbuffered location block below.
```

jgehrcke commented on packages/adminrouter/extra/src/test-harness/tests/test_service.py:
```diff
@@ -897,3 +897,81 @@ def test_if_requrl_rewriting_can_be_configured(
+        # request has been uploaded (Nginx buffers the data). Such a feature
+        # would introduce some extra complexity into the test harness. Simply
+        # checking if AR is printing the warning to the error log seems to be
+        # good enough.
```
I agree with that. Thanks for the code comment!!

vespian commented:
@mesosphere-mergebot bump-ee

adamtheturtle commented:
Please can you find the relevant `teamcity/dcos/test/docker` issue, and make a judgement call on whether this test suite needs to be seen passing for this PR.

vespian commented:
`teamcity/dcos/test/docker` CI failure seems like CI infra issue unrelated to the PR. I asked #tools-infra team and in the meantime, I have re-run the CI.

vespian commented:
@orsenthil  Could you please check why @mergebot still thinks that there are <2 reviews? Thanks!

vespian commented:
@orsenthil  It worked after @sambatyon re-approving the PR. I do not know why it is so, as there were no changes since the last time he approved it :(

orsenthil commented:
@vespian / @jgehrcke  - I am not sure why these are not counted. If you notice this, add a note to this JIRA -  https://jira.mesosphere.com/browse/QUALITY-1967
